File: plugin/skills/typescript/exampleDataFilter.ts
/**
 * Skill: Filter and transform data efficiently in execution environment
 * 
 * Demonstrates the key pattern: process large datasets HERE,
 * not through the LLM context window.
 */

export interface FilterOptions<T> {
  filterKey: keyof T;
  filterValue: any;
  outputFields?: (keyof T)[];
}

/**
 * Filter large dataset and extract only needed fields.
 */
export async function filterLargeDataset<T extends Record<string, any>>(
  data: T[],
  options: FilterOptions<T>
): Promise<Partial<T>[]> {
  const { filterKey, filterValue, outputFields } = options;
  
  let filtered = data.filter(row => row[filterKey] === filterValue);
  
  if (outputFields) {
    filtered = filtered.map(row => {
      const result: Partial<T> = {};
      for (const field of outputFields) {
        if (field in row) {
          result[field] = row[field];
        }
      }
      return result;
    });
  }
  
  console.log(`Filtered ${data.length} rows → ${filtered.length} rows`);
  return filtered;
}

File: plugin/skills/python/example_data_filter.py
"""
Skill: Filter and transform data efficiently in execution environment

Demonstrates the key pattern: process large datasets HERE,
not through the LLM context window.
"""

from typing import Any, Dict, List


async def filter_large_dataset(
    data: List[Dict[str, Any]], 
    filter_key: str,
    filter_value: Any,
    output_fields: List[str] = None
) -> List[Dict[str, Any]]:
    """
    Filter large dataset and extract only needed fields.
    
    Args:
        data: Large dataset to filter
        filter_key: Field to filter on
        filter_value: Value to match
        output_fields: Fields to keep (None = keep all)
    
    Returns:
        Filtered and transformed dataset
    """
    filtered = [row for row in data if row.get(filter_key) == filter_value]
    
    if output_fields:
        filtered = [
            {k: row[k] for k in output_fields if k in row}
            for row in filtered
        ]
    
    print(f"Filtered {len(data)} rows → {len(filtered)} rows")
    return filtered
